Improve contributing and testing documentation (#155)
queens-py/queens: CONTRIBUTING.md

# :busts_in_silhouette: Contributing to QUEENS
Thank you very much for your willingness to contribute to QUEENS! We strongly believe in the synergy effect of developing and using QUEENS together as a community.

> We invite you to share your methodological contributions that focus on probabilistic models and analyses, such as:

Reviewer: Maybe we could make it a little clearer that we welcome contributions to both deterministic and probabilistic methods, since the list that follows includes both method types:
```suggestion
We invite you to share your contributions to both deterministic and probabilistic methods, such as:
```
What do you think?

Author: I think it's strange since probabilistic methods strongly indicate that it is related to uq, ia, and so on. In contrast, deterministic seems a bit detached. But I don't mind using your description.

However, I do think it's important to add the 'probabilistic model' side, i.e., stuff like GP, normalizing flows, etc

Reviewer: That makes sense, maybe something like this fits better? Or would you like to emphasize the probabilistic aspect more?
```suggestion
We invite you to share your methodological contributions to both deterministic and probabilistic models and analyses, such as:
```

- Parameter studies and identification
- Sensitivity analysis
- Surrogate modeling
- Uncertainty quantification
- Bayesian inverse analysis
- Optimization

In addition to methodological contributions, we also greatly appreciate infrastructure contributions, which are essential to ensure that QUEENS runs smoothly and efficiently. In particular, but not limited to, we welcome bug fixes and improvements regarding:
- Code quality
- Performance
- User interface
- Benchmarking
- Testing

> - Tutorials

Reviewer: A link that guides the reader to the respective relevant testing and tutorials websites could be nice here

- [Documentation](#book-documentation)

We welcome all types of code contributions, irrespective of size and complexity.

> > Note: If you're unsure if your contribution fits into the QUEENS spirit feel free to ask by starting a [discussion](https://github.com/queens-py/queens/discussions) or opening an [issue](https://github.com/queens-py/queens/issues) :blush:

Reviewer: Since we typically use medium formal American English:
```suggestion
> Note: If you are unsure if your contribution fits into the QUEENS spirit, feel free to ask by starting a [discussion](https://github.com/queens-py/queens/discussions) or opening an [issue](https://github.com/queens-py/queens/issues) :blush:
```

> ### Contributing on GitHub
> - [Issues](#rotating_light-Issues)
> - [Pull requests](#fishing_pole_and_fish-pull-requests)

Reviewer: With the new structure, maybe we could remove this list altogether and move the "Issues" and "Pull request" headings on level down?
```suggestion
## Contributing on GitHub
```

# :busts_in_silhouette: Contributing to QUEENS
Thank you very much for your willingness to contribute to QUEENS! We strongly believe in the synergy effect of developing and using QUEENS together as a community.

We invite you to share your methodological contributions to both deterministic and probabilistic models and analyses, such as:
- Parameter studies and identification
- Sensitivity analysis
- Surrogate modeling
- Uncertainty quantification
- Bayesian inverse analysis
- Optimization

> In addition to methodological contributions, we also greatly appreciate infrastructure contributions, which are essential to ensure that QUEENS runs smoothly and efficiently. In particular, but not limited to, we welcome bug fixes and improvements regarding:

Reviewer 2: ```suggestion
In addition to methodological contributions, we also greatly appreciate infrastructure contributions that help ensure QUEENS runs smoothly and efficiently. These include, but are not limited to, bug fixes and improvements related to:
```

- Code quality
- Performance
- User interface
- Benchmarking
- [Testing](tests/README.md)
- [Tutorials](https://queens-py.github.io/queens/tutorials.html)
- [Documentation](doc/README.md)

We welcome all types of code contributions, irrespective of size and complexity.

> > Note: If you're unsure whether your contribution fits within the QUEENS framework, don't hesitate to ask the community by starting a [discussion](https://github.com/queens-py/queens/discussions) or by opening an [issue](https://github.com/queens-py/queens/issues) :blush:

Reviewer: I really like the way it sounds now, thank you! :relaxed:

# :busts_in_silhouette: Contributing to QUEENS
Thank you very much for your willingness to contribute to QUEENS! We strongly believe in the synergy effect of developing and using QUEENS together as a community.

We invite you to share your methodological contributions that focus on probabilistic models and analyses, such as:
- Parameter studies and identification
- Sensitivity analysis
- Surrogate modeling
- Uncertainty quantification
- Bayesian inverse analysis
- Optimization

In addition to methodological contributions, we also greatly appreciate infrastructure contributions, which are essential to ensure that QUEENS runs smoothly and efficiently. In particular, but not limited to, we welcome bug fixes and improvements regarding:
- Code quality
- Performance
- User interface
- Benchmarking
- [Testing](tests/README.md)
- [Tutorials](https://queens-py.github.io/queens/tutorials.html)

> - [Documentation](#book-documentation)

Author: @leahaeusel should we move the small documentation paragraph at the end of this file to the documentation readme and link to it instead? I think it fits better in the new structure.

Complete and meaningful docstrings are required as they are used to generate the
[documentation](#book-documentation).

##### QUEENS coding conventions

> As every codebase, we have some project-specific coding conventions. Here a list of common ones:

Reviewer 2: ```suggestion
Like every codebase, QUEENS follows some project-specific coding conventions. Below is a list of common ones:
```

- Use `pathlib.Path` objects instead of strings to handle paths and directories.
- If relative paths within the QUEENS source are needed, use the [relative_path_from_source](queens/utils/path.py#L23) function.
- Decorate the init method of QUEENS objects with the `log_init_args` decorator from [queens/utils/logger_setting.py](queens/utils/logger_settings.py#L239). This automatically logs the arguments passed to the init.

> - We only allow disabling pylint warnings for specific lines, not for entire files. If you disable warnings, please use the long pylint description.

Reviewer 2: ```suggestion
- We only allow disabling pylint warnings for specific lines, not for entire files. If you disable warnings, please use the long pylint description, not just the code.
```

Complete and meaningful docstrings are required as they are used to generate the
[documentation](#book-documentation).

#### QUEENS coding conventions
As every codebase, we have some project-specific coding conventions. Here a list of common ones:

> - Use `pathlib.Path` objects instead of string to handle paths and directories.

Reviewer: ```suggestion
- Use `pathlib.Path` objects instead of strings to handle paths and directories.
```

- If relative paths within the QUEENS source are needed, use the [relative_path_from_source](queens/utils/path.py#L23) function.

> - Decorate the init method of QUEENS-objects with the `log_init_args` decorator from [queens/utils/logger_setting.py](queens/utils/logger_settings.py#L239). This automatically logs the arguments passed to the init.

Reviewer: ```suggestion
- Decorate the init method of QUEENS objects with the `log_init_args` decorator from [queens/utils/logger_setting.py](queens/utils/logger_settings.py#L239). This automatically logs the arguments passed to the init.
```

> - We only allow to disable pylint warnings for specific lines, not for entire files. If you disable warnings please use the long pylint description.

Reviewer: ```suggestion
- We only allow disabling pylint warnings for specific lines, not for entire files. If you disable warnings, please use the long pylint description.
```
